### PR TITLE
Add support for chained gets, sets, and methods

### LINKED ISSUE #29: Nested accessors and methods
Currently when the rewriter encounters nested accessors and methods it will only convert the outermost to its original:

```
const pathLength = url.pathname.length;
const expressionName = node.getExpression().getSymbol().getName();

---- REWRITTEN ----

const pathLength = Reflect_apply(String_length_get, url.pathName);
const expressionName = Reflect_apply(Symbol_getName, node.getExpression().getSymbol(), []);
```

This is possibly because parents of observed nodes are being rewritten, which prevents the algorithm from continuing to inspect and properly rewrite all the children.

### PR DESCRIPTION
**Pull request by jackbsteinberg**
Resolves #29

### CODE REVIEW

File: lib/mods/replace-get-set.js [hunk 1]
```diff
@@ -48,7 +48,8 @@ function isDataPropertyAccess(node) {
  */
 function isGlobalPropertyAccess(node) {
   return nodeTypes.isPropertyAccessExpression(node) &&
-    !nodeTypes.isCallExpression(node.getParent()) &&
+    !(nodeTypes.isCallExpression(node.getParent()) &&
+    node.getParent().getExpression() === node) &&
```
Comment (domenic) [inline lib/mods/replace-get-set.js]: Is it possible to express what this condition is asking? e.g.

```js
const ___something___ = nodeTYpes.isCallExpression(node.getParent()) && node.getParent().getExpression === node;
```

and then just use `!___something__`? Where `__something__` is a semantic name indicating what actually you're testing for...

Comment (jackbsteinberg) [reply lib/mods/replace-get-set.js]: I was considering documenting via comments but this seems like a good way to do it, I'll change to this

**Comment by jackbsteinberg** [comment]
We only need the recursive application for methods and gets/sets, since, as of a few PRs ago, constructors don't rewrite. It's also unnecessary for the remove console mod as that only rewrites to `undefined` or deletes itself.

File: lib/mods/replace-get-set.js (hunk 1 above)
Comment (jackbsteinberg) [reply lib/mods/replace-get-set.js]: Updated to a separate var with a explanatory comment

**Comment by fergald** [comment]
> We only need the recursive application for methods and gets/sets, since, as of a few PRs ago, constructors don't rewrite. It's also unnecessary for the remove console mod as that only rewrites to `undefined` or deletes itself.

I still haven't figured out how this change works. If recursion only happens within a module, how does

```js
document.body.querySelector('class');
```

get rewritten correctly, that seems like we have to recurse into 2 modules.

Maybe I don't understand how the tree traversal works at all.

**Comment by jackbsteinberg** [comment]
I did some digging and that particular example works, but there is still an issue.

```js
document.body.querySelector('class');
```

This would first have the `replaceMethods` mod run, which would transform it into:

```js
Reflect_apply(HTMLElement_querySelector, document.body, ['class']
```

After the `replaceMethods` mod recursed and found no more methods to replace, it would have the `replaceGetsSets` mod run, which would replace `document.body` giving this:

```js
Reflect_apply(HTMLElement_querySelector, Reflect_apply(Document_body_get, document), ['class']);
```

After this `replaceGetsSets` would recurse, replacing `document` and leaving us with the final result of:

```js
Reflect_apply(HTMLElement_querySelector, Reflect_apply(Document_body_get, Window_document_get()), ['class']);
```

**Comment by jackbsteinberg** [comment]
This example works, but only because of the order of the accesses, with method coming before gets/sets. If, for example, we had something closer to:

```js
const otherDiff = document.querySelector('body').textContent;
```

Then the `textContent` portion of the statement would not be rewritten, leaving:

```js
Reflect_apply(Document_querySelector, Window_document_get(), ['body']).textContent;
```

This is an issue, and will likely require rethinking the strategy of only recursing with the current method.

**Pull request merged**